data_structures/bloom_filter.cpp

```diff
@@ -0,0 +1,288 @@
+ * to get hash for string
+ *
+ * @param s string to get hash from
+ * @return hash for a string
+ */
+static std::size_t hashStr(std::string const& s) {
+    std::size_t hash = 37;
+    std::size_t primeNum1 = 54059;
+    std::size_t primeNum2 = 76963;
+    for (char c : s) {
+        hash = (hash * primeNum1) ^ (c * primeNum2);
+    }
+    return hash;
+}
+
+/**
+ * @brief [Hash function for
+ * test](https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key)
+ *
+ * @param x int to get hash from
+ * @return hash for x
+ */
+std::size_t hashInt_1(int x) {
+    x = ((x >> 16) ^ x) * 0x45d9f3b;
+    x = ((x >> 16) ^ x) * 0x45d9f3b;
+    x = (x >> 16) ^ x;
+    return x;
+}
+
+/**
+ * @brief [Hash function for
+ * test](https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key)
+ *
+ * @param x int to get hash from
+ * @return hash for x
+ */
+std::size_t hashInt_2(int x) {
+    auto y = static_cast<std::size_t>(x);
+    y = (y ^ (y >> 30)) * static_cast<std::size_t>(0xbf58476d1ce4e5b9);
+    y = (y ^ (y >> 27)) * static_cast<std::size_t>(0x94d049bb133111eb);
+    y = y ^ (y >> 31);
+    return y;
+}
+} // namespace data_structures
+
+/**
+ * @brief Test for bloom filter with string as generic type
+ * @return void
+ */
+static void test_bloom_filter_string() {
+    data_structures::BloomFilter<std::string> filter(
+        10, {data_structures::hashDJB2, data_structures::hashStr});
+    std::vector<std::string> toCheck{"hello", "world", "!"};
+    std::vector<std::string> toFalse{"false", "world2", "!!!"};
+    for (auto& x : toCheck) {
+        filter.add(x);
+    }
+    for (auto& x : toFalse) {
+        assert(filter.contains(x) == false);
+    }
+    for (auto& x : toCheck) {
+        assert(filter.contains(x));
+    }
+}
+
+/**
+ * @brief Test for bloom filter with int as generic type
+ * @return void
+ */
+static void test_bloom_filter_int() {
+    data_structures::BloomFilter<int> filter(
+        20, {data_structures::hashInt_1, data_structures::hashInt_2});
+    std::vector<int> toCheck{100, 200, 300, 50};
+    std::vector<int> toFalse{1, 2, 3, 4, 5, 6, 7, 8};
+    for (int x : toCheck) {
+        filter.add(x);
+    }
+    for (int x : toFalse) {
+        assert(filter.contains(x) == false);
+    }
+    for (int x : toCheck) {
+        assert(filter.contains(x));
+    }
+}
+
+/**
+ * @brief Test for bitset
+ *
+ * @return void
+ */
+static void test_bitset() {
+    data_structures::Bitset set(2);
+    std::vector<std::size_t> toCheck{0, 1, 5, 8, 63, 64, 67, 127};
+    for (auto x : toCheck) {
+        set.add(x);
+        assert(set.contains(x));
+    }
+    assert(set.contains(128) == false);
+    assert(set.contains(256) == false);
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test_bitset();  // run test for bitset, because bloom filter is depend on it
```

Comment:
```suggestion
    // run self-test implementations
    test_bitset();  // run test for bitset, because bloom filter is depending on it
```


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix